### PR TITLE
[CBRD-20523] Foreign Key Referential Constraints check on INSERT/UPDATE

### PR DESCRIPTION
**Pull request by kisoo-han**
http://jira.cubrid.org/browse/CBRD-20523

TestCase for this change will be loaded into the repository 'cubrid-testcases'
sql/_30_banana_pie_qa/issue_20523_foreign_key/

### CODE REVIEW

File: src/transaction/locator_sr.c [hunk 1]
```diff
@@ -4201,7 +4201,8 @@ locator_check_foreign_key (THREAD_ENTRY * thread_p, HFID * hfid, OID * class_oid
 
       if (index->n_atts > 1)
 	{
-	  is_null = btree_multicol_key_is_null (key_dbvalue);
+	  /* If any column contains NULL value, skip foreign key constrains validity check */
+	  is_null = btree_multicol_key_has_null (key_dbvalue);
```
Comment (eseokoh) [inline src/transaction/locator_sr.c]: I'd like to propose to rename `is_null` as `has_null`.

File: src/transaction/locator_sr.c [hunk 2]
```diff
@@ -4201,7 +4201,8 @@ locator_check_foreign_key (THREAD_ENTRY * thread_p, HFID * hfid, OID * class_oid
 
       if (index->n_atts > 1)
 	{
-	  is_null = btree_multicol_key_is_null (key_dbvalue);
+	  /* If any column contains NULL value, skip foreign key constrains validity check */
```
Comment (eseokoh) [inline src/transaction/locator_sr.c]: Please change the comment line as follows:

```C
      /* SQL standard defines as follows:
       * If no <match type> was specified then, for each row R1 of the referencing table, either at least one of 
       * the values of the referencing columns in R1 shall be a null value, or the value of each referencing column 
       * in R1 shall be equal to the value of the corresponding referenced column in some row of the referenced table.
       * Please notice that we don't currently support <match type>. 
       */
      if (index->n_atts > 1)
```

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/transaction/locator_sr.c (6 incorrect indentation(s) found)
```
  4205 	   * If no <match type> was specified then, for each row R1 of the referencing table, 
  4206 	   * either at least one of the values of the referencing columns in R1 shall be a null value,
  4207 	   * or the value of each referencing column in R1 shall be equal to the value of 
  4208 	   * the corresponding referenced column in some row of the referenced table.
  4209 	   * Please notice that we don't currently support <match type>. 
  4210 	   */
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by eseokoh** [comment]
Would you please move the comment to out side of `if` block?
Since it explains two cases to satisfy a referential integrity, it is better to locate it on top of `if`.
In fact, my first remark shows it. Please have a look into it. :)

**Comment by eseokoh** [comment]
Is there a regression test case/expected answer requires to be modified?
Would you also open PR to test cases? I'd like to accept the patch as well as the test cases.

**Comment by kisoo-han** [comment]
It is not required to modify existing CUBRID test cases due to this modification[CBRD-20523].
We found same daily QA results before and after this changes were made.

For the new test case to test behavior of INSERT operation with composite foreign keys are added
and made pull request in GitHub(cubrid-testcases/sql/30bananapieqa/issue20523foreign_key/)

--
한기수 (연구소/기술위원)
Tel 070-4367-5140  Fax 070-8650-2150  Mobile 010-5213-2570
Email kisoo-han@cubrid.com  Facebook http://www.facebook.com/CUBRID.kr
[cid:B36A5B48-1412-4CDD-A081-A887E6977F2D]

**Comment by eseokoh** [comment]
Sorry I'm unable to find the PR. 
Would you let us know the link to the PR for the test cases?

**Comment by kisoo-han** [comment]
I made a pull request on GitHub curbed-testcases, following is the link:

https://github.com/kisoo-han/cubrid-testcases/tree/master/sql/_30_banana_pie_qa/issue_20523_foreign_key

--
한기수 (연구소/기술위원)
Tel 070-4367-5140  Fax 070-8650-2150  Mobile 010-5213-2570
Email kisoo-han@cubrid.com  Facebook http://www.facebook.com/CUBRID.kr
[cid:852E9762-CAE5-422E-962F-37C17BCED8E6]

**Comment by eseokoh** [comment]
ah. It is in your own fork. 
Please issue a PR to upstream: https://github.com/CUBRID/cubrid-testcases
Since the existing cases does not cover the cases, we surely need regression test cases as well as the patch. 

Another ask is to add test cases to cover `UPDATE` case.
Please notice that `locator_check_foreign_key` can also be reached from `locator_update_force`.

**Comment by kisoo-han** [comment]
One more test case is added to check UPDATE case, and made a Pull Request.
Please find GitHub curbrid-testcases and if you have a problem to find test cases, let me know it.

--
한기수 (연구소/기술위원)
Tel 070-4367-5140  Fax 070-8650-2150  Mobile 010-5213-2570
Email kisoo-han@cubrid.com  Facebook http://www.facebook.com/CUBRID.kr
[cid:2CE190E7-CE8E-4F68-B601-58F0B92A6416]

**Comment by eseokoh** [comment]
Please refer to https://github.com/CUBRID/cubrid/wiki#merge-pull-request and do "Squash and merge" your changes including test cases of https://github.com/CUBRID/cubrid-testcases/pull/36.

Thanks for your first contribution. :)

**Pull request merged**